src/libraries/System.Net.Primitives/src/System/Net/IPNetwork.cs

```diff
@@ -0,0 +1,344 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Net.Sockets;
+
+namespace System.Net
+{
+    /// <summary>
+    /// Provides an Internet Protocol (IP) subnet/range in CIDR notation.
+    /// </summary>
+    public readonly struct IPNetwork : IEquatable<IPNetwork>, ISpanFormattable, ISpanParsable<IPNetwork>
+    {
+        public IPAddress BaseAddress { get; private init; }
+        public int PrefixLength { get; private init; }
+
+        /*
+         We can also add two more easy to implement properties:
+         * AddressCount - a lazily calculated property showing the amount of potential address this instance "Contains". 2^(128_OR_32 - PrefixLength)
+            * 2^128 is going to fit into BigInteger, which causes a concern for memory allocation, since for actually big integers BigInteger seems to allocate a byte array.
+            * We can just have a Nullable<BigInteger> backing field for the lazy property. But we should be fine with the increased byte size for IPNetwork type itself caused by that.
+         * Since BaseAddress represents the first address in the range, does it make sense to have a property that represents the last address?
+         */
+
+        private const char addressAndPrefixLengthSeparator = '/';
+        private const int bitsPerByte = 8;
+        private const string
+            baseAddressConstructorParamName = "baseAddress",
+            prefixLengthConstructorParamName = "prefixLength";
+        public IPNetwork(IPAddress baseAddress, int prefixLength)
+        {
+            BaseAddress = baseAddress;
+            PrefixLength = prefixLength;
+
+            var validationError = Validate();
```

Comment:
Unless the type is state on the right-hand side, please use the explicit type name rather than var.